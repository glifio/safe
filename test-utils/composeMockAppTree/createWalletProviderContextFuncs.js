import { FilecoinNumber } from '@glif/filecoin-number'
import { createPath } from '@glif/react-components'
import { TESTNET_PATH_CODE } from '../../constants'

import {
  setLoginOption,
  setError,
  resetLedgerState,
  resetState,
  walletList,
  updateBalance
} from '@glif/react-components'
import { mockWalletProviderInstance } from '../../__mocks__/@glif/filecoin-wallet-provider'
import { WALLET_ADDRESS_2, WALLET_ID_2 } from '../constants'

export const mockFetchDefaultWallet = jest.fn().mockImplementation(() => ({
  balance: new FilecoinNumber('1', 'fil'),
  robust: WALLET_ADDRESS_2,
  id: WALLET_ID_2,
  path: createPath(TESTNET_PATH_CODE, 0)
}))

export default (walletProviderDispatch, state) => ({
  fetchDefaultWallet: mockFetchDefaultWallet,
  walletList: jest
    .fn()
    .mockImplementation((wallets) =>
      walletProviderDispatch(walletList(wallets))
    ),
  setWalletError: jest
    .fn()
    .mockImplementation((errorMessage) =>
      walletProviderDispatch(setError(errorMessage))
    ),
  setLoginOption: jest
    .fn()
    .mockImplementation((loginOption) =>
      walletProviderDispatch(setLoginOption(loginOption))
    ),
  connectLedger: jest.fn().mockImplementation(() => mockWalletProviderInstance),
  connectMetaMask: jest.fn().mockImplementation(() => {
    walletProviderDispatch({ type: 'METAMASK_CONFIGURED_SUCCESS' })
    return mockWalletProviderInstance
  }),
  getProvider: jest.fn().mockImplementation(() => mockWalletProviderInstance),
  resetLedgerState: jest.fn().mockImplementation(() => {
    walletProviderDispatch(resetLedgerState())
  }),
  resetState: jest.fn().mockImplementation(() => {
    walletProviderDispatch(resetState())
  }),
  updateBalance: jest.fn().mockImplementation((balance, index) => {
    walletProviderDispatch(updateBalance(balance, index))
  }),
  walletError: jest.fn().mockImplementation(() => state.error),
  resetWalletError: jest.fn().mockImplementation(() => {})
})
