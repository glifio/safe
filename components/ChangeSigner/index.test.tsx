import {
  render,
  act,
  waitFor,
  fireEvent,
  getByText,
  getByRole,
  RenderResult
} from '@testing-library/react'
import { Context } from 'react'
import { BigNumber } from '@glif/filecoin-number'
import { Message } from '@glif/filecoin-message'
import {
  MsigMethod,
  truncateAddress,
  WalletProviderContextType
} from '@glif/react-components'

import {
  pushPendingMessageSpy,
  WalletProviderContext,
  PendingMsgContext
} from '../../__mocks__/@glif/react-components'
import composeMockAppTree from '../../test-utils/composeMockAppTree'
import {
  WALLET_ADDRESS,
  MULTISIG_ACTOR_ADDRESS,
  MULTISIG_SIGNER_ADDRESS_2
} from '../../test-utils/constants'
import { ChangeSigner } from '.'

const newAddress = 't1iuryu3ke2hewrcxp4ezhmr5cmfeq3wjhpxaucza'

jest.mock('@glif/filecoin-wallet-provider')

describe('ChangeSigner', () => {
  test('it allows a user to change a signer address', async () => {
    const { Tree, walletProvider } = composeMockAppTree('postOnboard')
    let result: RenderResult | null = null

    await act(async () => {
      result = render(
        <Tree>
          <ChangeSigner
            oldSignerAddress={MULTISIG_SIGNER_ADDRESS_2}
            walletProviderOpts={{
              context:
                WalletProviderContext as unknown as Context<WalletProviderContextType>
            }}
            pendingMsgContext={PendingMsgContext}
          />
        </Tree>
      )

      // Get HTML elements
      const header = getByRole(result.container, 'heading')
      const oldSigner = getByRole(result.container, 'combobox')
      const newSigner = getByRole(result.container, 'textbox')
      const cancel = getByText(result.container, 'Cancel')
      const review = getByText(result.container, 'Review')

      // Check initial state
      expect(header).toHaveTextContent('Change a signer')
      expect(oldSigner).toHaveDisplayValue(
        truncateAddress(MULTISIG_SIGNER_ADDRESS_2)
      )
      expect(newSigner).toHaveFocus()
      expect(newSigner).toHaveDisplayValue('')
      expect(cancel).toBeEnabled()
      expect(review).toBeDisabled()

      // Enter new address
      newSigner.focus()
      fireEvent.change(newSigner, { target: { value: newAddress } })
      newSigner.blur()
      jest.runAllTimers()

      // Review should now be enabled
      expect(review).toBeEnabled()

      // Click review
      fireEvent.click(review)
      jest.runAllTimers()

      // The total amount should show after getting the tx fee
      await waitFor(
        () => expect(getByText(result.container, 'Total')).toBeInTheDocument(),
        { timeout: 10000 }
      )

      // The tx fee info should now be shown
      const maxFeeRegex =
        /You will not pay more than [0-9.]+ FIL for this transaction/i
      expect(getByText(result.container, maxFeeRegex)).toBeInTheDocument()

      // The expert mode toggle should shown and be off
      const expertMode = getByRole(result.container, 'checkbox')
      expect(expertMode).toBeInTheDocument()
      expect(expertMode).not.toBeChecked()

      // The send button should now be available
      const send = getByText(result.container, 'Send')
      expect(send).toBeInTheDocument()
      expect(send).toBeEnabled()

      // Click send
      fireEvent.click(send)
      jest.runAllTimers()
    })

    // Check wallet provider calls
    expect(walletProvider.getNonce).toHaveBeenCalled()
    expect(walletProvider.wallet.sign).toHaveBeenCalled()
    expect(walletProvider.simulateMessage).toHaveBeenCalled()
    expect(walletProvider.sendMessage).toHaveBeenCalled()

    // Check if sent message was properly formatted
    const lotusMessage = walletProvider.wallet.sign.mock.calls[0][1]
    const message = Message.fromLotusType(lotusMessage)
    expect(message.from).toBe(WALLET_ADDRESS)
    expect(message.to).toBe(MULTISIG_ACTOR_ADDRESS)
    expect(message.nonce).toBeGreaterThanOrEqual(0)
    expect(BigNumber.isBigNumber(message.value)).toBe(true)
    expect(message.value.isEqualTo(0)).toBe(true)
    expect(message.method).toBe(MsigMethod.PROPOSE)
    expect(typeof message.params).toBe('string')
    expect(message.params).toBeTruthy()
    expect(BigNumber.isBigNumber(message.gasPremium)).toBe(true)
    expect(message.gasPremium.isGreaterThan(0)).toBe(true)
    expect(BigNumber.isBigNumber(message.gasFeeCap)).toBe(true)
    expect(message.gasFeeCap.isGreaterThan(0)).toBe(true)
    expect(message.gasLimit).toBeGreaterThan(0)

    // Check if pending message was properly pushed
    const pendingMsg = pushPendingMessageSpy.mock.calls[0][0]
    expect(typeof pendingMsg.cid).toBe('string')
    expect(pendingMsg.cid).toBeTruthy()
    expect(pendingMsg.from.robust).toBe(WALLET_ADDRESS)
    expect(pendingMsg.to.robust).toBe(MULTISIG_ACTOR_ADDRESS)
    expect(pendingMsg.height).toBe(0)
    expect(typeof pendingMsg.params).toBe('string')
    expect(pendingMsg.params).toBeTruthy()
    expect(Number(pendingMsg.nonce)).toBeGreaterThanOrEqual(0)
    expect(Number(pendingMsg.method)).toBe(MsigMethod.PROPOSE)
    expect(Number(pendingMsg.value)).toBe(0)
    expect(Number(pendingMsg.gasFeeCap)).toBeGreaterThan(0)
    expect(Number(pendingMsg.gasLimit)).toBeGreaterThan(0)
    expect(Number(pendingMsg.gasPremium)).toBeGreaterThan(0)

    // Check snapshot
    expect(result.container.firstChild).toMatchSnapshot()
  })

  test('it renders the initial state correctly', async () => {
    const { Tree } = composeMockAppTree('postOnboard')
    let result: RenderResult | null = null

    await act(async () => {
      result = render(
        <Tree>
          <ChangeSigner oldSignerAddress={MULTISIG_SIGNER_ADDRESS_2} />
        </Tree>
      )
    })

    // Check snapshot
    expect(result.container.firstChild).toMatchSnapshot()
  })
})
