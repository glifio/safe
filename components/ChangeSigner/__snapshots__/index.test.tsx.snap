// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChangeSigner it allows a user to change a signer address 1`] = `
.c16 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  white-space: nowrap;
  cursor: pointer;
  gap: 0.75rem;
  padding: 0.5em 0.75em;
  color: var(--black);
  border-width: 1px;
  border-style: solid;
  border-radius: 8px;
  border-color: var(--black);
  background: transparent;
  font-size: 1.25em;
  color: var(--white);
  border-color: var(--green-medium);
  background: var(--green-medium);
  pointer-events: none;
  color: var(--gray-medium);
  border-color: var(--gray-medium);
  background: transparent;
}

.c16 svg path {
  stroke: var(--black);
}

.c16:hover {
  color: var(--purple-medium);
  border-color: var(--purple-medium);
}

.c16:hover svg path {
  stroke: var(--purple-medium);
}

.c16:active {
  color: var(--white) !important;
  border-color: var(--purple-medium) !important;
  background: var(--purple-medium) !important;
}

.c16:active svg path {
  stroke: var(--white) !important;
}

.c16 svg path {
  stroke: var(--white);
}

.c16:hover {
  color: var(--green-medium);
  border-color: var(--green-medium);
  background: var(--white);
}

.c16:hover svg path {
  stroke: var(--green-medium);
}

.c16 svg path {
  stroke: var(--gray-medium);
}

.c15 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  white-space: nowrap;
  cursor: pointer;
  gap: 0.75rem;
  padding: 0.5em 0.75em;
  color: var(--black);
  border-width: 1px;
  border-style: solid;
  border-radius: 8px;
  border-color: var(--black);
  background: transparent;
  font-size: 1.25em;
  pointer-events: none;
  color: var(--gray-medium);
  border-color: var(--gray-medium);
  background: transparent;
}

.c15 svg path {
  stroke: var(--black);
}

.c15:hover {
  color: var(--purple-medium);
  border-color: var(--purple-medium);
}

.c15:hover svg path {
  stroke: var(--purple-medium);
}

.c15:active {
  color: var(--white) !important;
  border-color: var(--purple-medium) !important;
  background: var(--purple-medium) !important;
}

.c15:active svg path {
  stroke: var(--white) !important;
}

.c15 svg path {
  stroke: var(--gray-medium);
}

.c3 {
  display: inline-block;
  height: 48px;
  width: 48px;
  text-align: left;
  background-color: var(--purple-medium);
}

.c4 {
  display: inline-block;
  height: 24px;
  width: 24px;
  background-color: var(--purple-dark);
  -webkit-animation-name: bvGDKI;
  animation-name: bvGDKI;
  -webkit-animation-duration: 2s;
  animation-duration: 2s;
  -webkit-animation-iteration-count: infinite;
  animation-iteration-count: infinite;
}

.c1 {
  padding: 1.5em;
  border-radius: 8px;
  overflow: hidden;
  text-align: center;
  word-break: break-word;
}

.c1 > *:first-child {
  margin-top: 0;
}

.c1 > *:last-child {
  margin-bottom: 0;
}

.c1 > svg:first-child {
  display: block;
  margin: 0 auto 1em;
}

.c1 > h2:first-child {
  margin: 0;
}

.c1 header:first-child {
  margin: -1.5em -1.5em 1.5em -1.5em;
  padding: 1.5em;
  background-color: var(--purple-medium);
  color: var(--white);
}

.c2 {
  background-color: var(--blue-gray);
}

.c2 hr {
  border-color: var(--gray-medium);
}

.c5 {
  background-color: var(--white);
  box-shadow: 0 0 0.5em var(--gray-light);
}

.c13 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 10px;
  margin: 1.5em 0;
}

.c14 {
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 10px;
  width: 100%;
  max-width: 35em;
}

.c0 form {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 10px;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 10px;
}

.c6 > p,
.c6 > hr {
  margin: 0.5em 0;
}

.c6 > p[role='button'] {
  margin: 0;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 10px;
  text-align: left;
  pointer-events: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c8 > *:first-child {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c8 > *:last-child:not(.toggle-wrapper) {
  -webkit-flex: 0 0 50%;
  -ms-flex: 0 0 50%;
  flex: 0 0 50%;
}

.c8 span.info,
.c8 span.error {
  font-size: 0.875rem;
}

.c8 span.error {
  color: var(--red-medium);
}

.c10 .button-wrapper {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 10px;
}

.c10 .button-wrapper > *:first-child {
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
}

.c10 .button-wrapper > *:not(:first-child) {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  -webkit-transition: -webkit-transform 0.1s ease-out;
  -webkit-transition: transform 0.1s ease-out;
  transition: transform 0.1s ease-out;
}

.c10 .button-wrapper > *:not(:first-child):hover:not(:active) {
  -webkit-transform: scale(1.2);
  -ms-transform: scale(1.2);
  transform: scale(1.2);
}

.c10 .button-wrapper > *:not(:first-child):active {
  -webkit-transition: none;
  transition: none;
}

.c10 .unit-wrapper {
  position: relative;
}

.c10 .unit-wrapper input {
  width: 100%;
}

.c10 .unit-wrapper .unit {
  position: absolute;
  top: 50%;
  right: 1em;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  color: var(--purple-medium);
  color: var(--gray-dark) !important;
}

.c9 .select-wrapper {
  position: relative;
}

.c9 .select-wrapper select {
  width: 100%;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c9 .select-wrapper .select-arrow {
  position: absolute;
  pointer-events: none;
  top: 50%;
  right: 1.2em;
  width: 0.6em;
  height: 0.6em;
  border-right: 1.5px solid var(--black);
  border-bottom: 1.5px solid var(--black);
  -webkit-transform: translateY(-0.35em) rotate(45deg);
  -ms-transform: translateY(-0.35em) rotate(45deg);
  transform: translateY(-0.35em) rotate(45deg);
  border-color: var(--gray-dark);
}

.c11 .toggle-wrapper {
  position: relative;
}

.c11 input {
  position: absolute;
  top: 50%;
  left: 50%;
  width: 0;
  height: 0;
  opacity: 0;
}

.c11 .toggle {
  display: inline-block;
  position: relative;
  width: 2.25em;
  height: 1.5em;
  border-radius: 0.75em;
  border: 2px solid var(--blue-medium);
  background-color: var(--blue-medium);
  border-color: var(--gray-light) !important;
  background-color: var(--gray-light) !important;
}

.c11 .toggle::after {
  content: '';
  position: absolute;
  top: 0;
  left: 0;
  width: calc(1.5em - 4px);
  height: calc(1.5em - 4px);
  border-radius: 50%;
  background-color: var(--white);
}

.c11 input:checked ~ .toggle {
  border-color: var(--blue-dark);
  background-color: var(--blue-dark);
}

.c11 input:checked ~ .toggle::after {
  left: auto;
  right: 0;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  gap: 20px;
}

.c7 > * {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
}

.c7 > *:first-child {
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
}

.c12 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c12 span {
  color: var(--purple-medium);
  font-size: 1.75rem;
}

<div
  class="c0"
>
  <form>
    <div
      class="c1 c2"
    >
      <h2>
        Change a signer
      </h2>
      <hr />
      <p>
        Sending your transaction...
      </p>
      <span
        class="c3"
      >
        <span
          class="c4"
        />
      </span>
    </div>
    
    <div
      class="c1 c5"
    >
      <div
        class="c6"
      >
        <header
          class="c7"
        >
          <div>
            <span>
              From
            </span>
            <span>
              t1z225 ... 6z6wgi
            </span>
          </div>
          <div>
            <span>
              Signer 
              Balance
            </span>
            <span>
              1
               FIL
            </span>
          </div>
          <div>
            <span>
              Safe Balance
            </span>
            <span>
              1
               FIL
            </span>
          </div>
        </header>
        <label
          class="c8 c9"
          disabled=""
        >
          <div>
            <span>
              Old Signer
            </span>
            
          </div>
          <div
            class="select-wrapper"
          >
            <select
              disabled=""
            >
              
              <option
                value="t1nq5k2mps5umtebdovlyo7y6a3ywc7u4tobtuo3a"
              >
                t1nq5k ... btuo3a
              </option>
            </select>
            <div
              class="select-arrow"
            />
          </div>
        </label>
        <label
          class="c8 c10"
          disabled=""
        >
          <div>
            <span>
              New Signer
            </span>
            
            
          </div>
          <div
            class="button-wrapper"
          >
            <div
              class="unit-wrapper"
            >
              <input
                autocomplete="on"
                class=""
                disabled=""
                max=""
                min=""
                name=""
                placeholder="f1..."
                step="any"
                style="padding-right: 1em;"
                type="text"
                value="t1iury ... xaucza"
              />
              
            </div>
          </div>
        </label>
        <hr />
        <label
          class="c8 c11"
          disabled=""
        >
          <div>
            <span>
              Expert Mode
            </span>
            
          </div>
          <div
            class="toggle-wrapper"
          >
            <input
              disabled=""
              type="checkbox"
            />
            <span
              class="toggle"
            />
          </div>
        </label>
        <p>
          You will not pay more than 
          0.00000000001
           FIL for this transaction.
           
          <a
            href="https://filfox.info/en/stats/gas"
            rel="noreferrer noopener"
            target="_blank"
          >
            More information on average gas fee statistics.
          </a>
        </p>
        <hr />
        <div
          class="c12"
        >
          <span>
            Total
          </span>
          <span>
            0.00000000001
             FIL
          </span>
        </div>
      </div>
    </div>
    <div
      class="c13 c14"
    >
      <button
        class="c15"
        disabled=""
        type="button"
      >
        Back
      </button>
      <button
        class="c16"
        disabled=""
        type="submit"
      >
        Send
      </button>
    </div>
  </form>
</div>
`;

exports[`ChangeSigner it renders the initial state correctly 1`] = `
.c11 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  white-space: nowrap;
  cursor: pointer;
  gap: 0.75rem;
  padding: 0.5em 0.75em;
  color: var(--black);
  border-width: 1px;
  border-style: solid;
  border-radius: 8px;
  border-color: var(--black);
  background: transparent;
  font-size: 1.25em;
}

.c11 svg path {
  stroke: var(--black);
}

.c11:hover {
  color: var(--purple-medium);
  border-color: var(--purple-medium);
}

.c11:hover svg path {
  stroke: var(--purple-medium);
}

.c11:active {
  color: var(--white) !important;
  border-color: var(--purple-medium) !important;
  background: var(--purple-medium) !important;
}

.c11:active svg path {
  stroke: var(--white) !important;
}

.c12 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  white-space: nowrap;
  cursor: pointer;
  gap: 0.75rem;
  padding: 0.5em 0.75em;
  color: var(--black);
  border-width: 1px;
  border-style: solid;
  border-radius: 8px;
  border-color: var(--black);
  background: transparent;
  font-size: 1.25em;
  color: var(--white);
  border-color: var(--green-medium);
  background: var(--green-medium);
  pointer-events: none;
  color: var(--gray-medium);
  border-color: var(--gray-medium);
  background: transparent;
}

.c12 svg path {
  stroke: var(--black);
}

.c12:hover {
  color: var(--purple-medium);
  border-color: var(--purple-medium);
}

.c12:hover svg path {
  stroke: var(--purple-medium);
}

.c12:active {
  color: var(--white) !important;
  border-color: var(--purple-medium) !important;
  background: var(--purple-medium) !important;
}

.c12:active svg path {
  stroke: var(--white) !important;
}

.c12 svg path {
  stroke: var(--white);
}

.c12:hover {
  color: var(--green-medium);
  border-color: var(--green-medium);
  background: var(--white);
}

.c12:hover svg path {
  stroke: var(--green-medium);
}

.c12 svg path {
  stroke: var(--gray-medium);
}

.c1 {
  padding: 1.5em;
  border-radius: 8px;
  overflow: hidden;
  text-align: center;
  word-break: break-word;
}

.c1 > *:first-child {
  margin-top: 0;
}

.c1 > *:last-child {
  margin-bottom: 0;
}

.c1 > svg:first-child {
  display: block;
  margin: 0 auto 1em;
}

.c1 > h2:first-child {
  margin: 0;
}

.c1 header:first-child {
  margin: -1.5em -1.5em 1.5em -1.5em;
  padding: 1.5em;
  background-color: var(--purple-medium);
  color: var(--white);
}

.c2 {
  background-color: var(--blue-gray);
}

.c2 hr {
  border-color: var(--gray-medium);
}

.c3 {
  background-color: var(--white);
  box-shadow: 0 0 0.5em var(--gray-light);
}

.c9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 10px;
  margin: 1.5em 0;
}

.c10 {
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 10px;
  width: 100%;
  max-width: 35em;
}

.c0 form {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 10px;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 10px;
}

.c4 > p,
.c4 > hr {
  margin: 0.5em 0;
}

.c4 > p[role='button'] {
  margin: 0;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 10px;
  text-align: left;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c6 > *:first-child {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c6 > *:last-child:not(.toggle-wrapper) {
  -webkit-flex: 0 0 50%;
  -ms-flex: 0 0 50%;
  flex: 0 0 50%;
}

.c6 span.info,
.c6 span.error {
  font-size: 0.875rem;
}

.c6 span.error {
  color: var(--red-medium);
}

.c8 .button-wrapper {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 10px;
}

.c8 .button-wrapper > *:first-child {
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
}

.c8 .button-wrapper > *:not(:first-child) {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  -webkit-transition: -webkit-transform 0.1s ease-out;
  -webkit-transition: transform 0.1s ease-out;
  transition: transform 0.1s ease-out;
}

.c8 .button-wrapper > *:not(:first-child):hover:not(:active) {
  -webkit-transform: scale(1.2);
  -ms-transform: scale(1.2);
  transform: scale(1.2);
}

.c8 .button-wrapper > *:not(:first-child):active {
  -webkit-transition: none;
  transition: none;
}

.c8 .unit-wrapper {
  position: relative;
}

.c8 .unit-wrapper input {
  width: 100%;
}

.c8 .unit-wrapper .unit {
  position: absolute;
  top: 50%;
  right: 1em;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  color: var(--purple-medium);
}

.c7 .select-wrapper {
  position: relative;
}

.c7 .select-wrapper select {
  width: 100%;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c7 .select-wrapper .select-arrow {
  position: absolute;
  pointer-events: none;
  top: 50%;
  right: 1.2em;
  width: 0.6em;
  height: 0.6em;
  border-right: 1.5px solid var(--black);
  border-bottom: 1.5px solid var(--black);
  -webkit-transform: translateY(-0.35em) rotate(45deg);
  -ms-transform: translateY(-0.35em) rotate(45deg);
  transform: translateY(-0.35em) rotate(45deg);
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  gap: 20px;
}

.c5 > * {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
}

.c5 > *:first-child {
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
}

<div
  class="c0"
>
  <form>
    <div
      class="c1 c2"
    >
      <h2>
        Change a signer
      </h2>
      <hr />
      <p>
        Please update the signer address below
      </p>
    </div>
    
    <div
      class="c1 c3"
    >
      <div
        class="c4"
      >
        <header
          class="c5"
        >
          <div>
            <span>
              From
            </span>
            <span>
              t1z225 ... 6z6wgi
            </span>
          </div>
          <div>
            <span>
              Signer 
              Balance
            </span>
            <span>
              1
               FIL
            </span>
          </div>
          <div>
            <span>
              Safe Balance
            </span>
            <span>
              1
               FIL
            </span>
          </div>
        </header>
        <label
          class="c6 c7"
        >
          <div>
            <span>
              Old Signer
            </span>
            
          </div>
          <div
            class="select-wrapper"
          >
            <select>
              
              <option
                value="t1nq5k2mps5umtebdovlyo7y6a3ywc7u4tobtuo3a"
              >
                t1nq5k ... btuo3a
              </option>
            </select>
            <div
              class="select-arrow"
            />
          </div>
        </label>
        <label
          class="c6 c8"
        >
          <div>
            <span>
              New Signer
            </span>
            
            
          </div>
          <div
            class="button-wrapper"
          >
            <div
              class="unit-wrapper"
            >
              <input
                autocomplete="on"
                class=""
                max=""
                min=""
                name=""
                placeholder="f1..."
                step="any"
                style="padding-right: 1em;"
                type="text"
                value=""
              />
              
            </div>
          </div>
        </label>
      </div>
    </div>
    <div
      class="c9 c10"
    >
      <button
        class="c11"
        type="button"
      >
        Cancel
      </button>
      <button
        class="c12"
        disabled=""
        type="submit"
      >
        Review
      </button>
    </div>
  </form>
</div>
`;
